src/core/namespace.c

```diff
@@ -1374,6 +1393,60 @@ static bool home_read_only(
         return false;
 }
 
+static int verity_settings_prepare(
+                VeritySettings *verity,
+                const char *root_image,
+                const void *root_hash,
+                size_t root_hash_size,
+                const char *root_hash_path,
+                const void *root_hash_sig,
+                size_t root_hash_sig_size,
+                const char *root_hash_sig_path,
+                const char *verity_data_path) {
+
+        int r;
+
+        assert(verity);
+
+        if (root_hash) {
+                void *d;
+
+                d = memdup(root_hash, root_hash_size);
+                if (!d)
+                        return -ENOMEM;
+
+                free_and_replace(verity->root_hash, d);
+                verity->root_hash_size = root_hash_size;
```

Comment:
Isn't the normal pattern to avoid side effects in case of errors, and modifying variables only on success?

Reply:
well, yes, and no.

For stuff that is initialized anew we pretty strictly insist on behaviour that either we initialize in full or not at all. But this case isn't like that. We augment an existing pre-initialized object here, add in a couple of other things. For those it's nice to have atomic behaviour too (i.e. fail entirely or succeed completely), but we are much less strict here, simply because it's sometimes really hard and often not useful. What we do insist on in cases like that: leave all resources around in a fashion that they are fully tracked and thus released if the object is gone, i.e. if a function consists of multiple steps (like this one, which basically  does three separate tjings) then make sure each one is atomic even if the combination of the three is not really. 

Hope this makes some sense?

Or to say this differently: on output-only params we are strict. on input/output params we are much less strict, we just want that everything is released when the param is eventually freed.

Reply:
sure, makes sense